spec.html

```diff
@@ -11059,6 +11060,26 @@ <h1>Examples of Automatic Semicolon Insertion</h1>
       <p>In the circumstance that an assignment statement must begin with a left parenthesis, it is a good idea for the programmer to provide an explicit semicolon at the end of the preceding statement rather than to rely on automatic semicolon insertion.</p>
     </emu-clause>
   </emu-clause>
+  <emu-clause id="sec-hazards-of-automatic-semicolon-insertion">
+    <h1>Hazards of Automatic Semicolon Insertion</h1>
+    <em>This section is non-normative.</em>
+    <p>ECMAScript programs can be written in a style with very few semicolons by relying on automatic semicolon insertion. However, as described above, semicolons are not inserted at every newline, and some of these cases which do not have an automatically inserted semicolon can be counter-intuitive or confusing.</p>
+
+    <p>As new syntactic features are added to ECMAScript, additional cases requiring explicit semicolons emerge over time. As such, consistently explicit semicolon use is recommended.</p>
```

Comment:
The placement of the recommendation doesn't seem ideal. It is true that as the grammar increases, new cases of explicit semicolon insertion will appear (the template tag one is the most relevant example), but making the recommendation based on speculations about the future implies that the committee is open to grammar changes that modify the semantics of currently valid javascript programs.

It would make more sense to place the recommendation at the end, as a pragmatic advice for avoiding ASI cases that are known to be confusing, without implying a lack of commitment to backwards compat from the part of the committee.

Reply:
this recommendation's discussion was brought up from a current proposal that is already being staged for browser usage and has heavy ecosystem reliance due to transpilers : https://tc39.github.io/proposal-class-fields/#sec-asi-hazards-in-class-bodies , https://github.com/rwaldron/tc39-notes/blob/master/es8/2017-11/nov-30.md#10ivf-class-fields-asi-discussion-and-resolution

Reply:
That's exactly the same case as template tags. Under ES5 it's impossible to write template strings, so it's impossible to have a program that behaves differently in ES5 and ES6 wrt a line break followed by a backtick. Likewise, it's impossible to write a program today that will behave differently once class fields land because the proposed grammar is currently illegal. So saying semicolons make existing code future-proof is an argument that is not supported by previous occurrences, and on top of that, it implies that there are plans to break existing semantics.

For every new feature that introduces a new ASI pitfall, you'll need to document the new case in that version of the spec anyways, and that pitfall will not affect previous versions of the spec. It makes the most sense to make the recommendation as a blanket advice for the cases that are known for each version of the spec, no more and no less.

Reply:
The hazard is not about existing code; it's about programmers using existing *patterns* with new features, without being aware of the hazard - which is precisely [the kind of thing that happened](https://github.com/eslint/eslint/issues/4210) with tagged template literals when they were introduced.

Reply:
That's good thing about ASI as it teaches you not to write weird things. Anyone who starts a line with a template literal (or an array) has greater problems about code style to fix than wage a war online over usage of a character.

Reply:
rephrased in more technical terms